mkosi/machine.py

```diff
@@ -0,0 +1,131 @@
+            tmp.network_veth = True
+            tmp.ssh = True
+        elif tmp.verb == "boot":
+            pass
+        else:
+            die("No valid verb was entered.")
+
+        self.args = load_args(tmp)
+
+
+    def ensure_booted(self) -> pexpect.spawn:
+        if self._serial is None:
+            raise ValueError(
+                        dedent(
+                            """\
+                            Trying to access serial console before machine boot or after machine shutdown.
+                            In order to boot the machine properly, use it as a context manager.
+                            Then, a Mkosi image will be booted in the __enter__ method.
+                            """
+                        )
+                    )
+        return self._serial
+
+
+    def __enter__(self) -> Machine:
+        if self.args.verb in MKOSI_COMMANDS_SUDO:
+            check_root()
+            unlink_output(self.args)
+
+        if self.args.verb == "build":
+            check_output(self.args)
+
+        if needs_build(self.args):
+            check_root()
+            check_native(self.args)
+            init_namespace(self.args)
+            build_stuff(self.args)
+
+        with contextlib.ExitStack() as stack:
+            prepend_to_environ_path(self.args.extra_search_paths)
+
+            if self.args.verb in ("shell", "boot"):
+                cmdline = run_shell_cmdline(self.args)
+            elif self.args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(run_qemu_cmdline(self.args))
+            else:
+                die("No valid verb was entered.")
+
+            cmd = " ".join(str(x) for x in cmdline)
+
+            # Here we have something equivalent to the command lines used on spawn() and run() from backend.py
+            # We use pexpect to boot an image that we will be able to interact with in the future
+            # Then we tell the process to look for the # sign, which indicates the CLI for that image is active
+            # Once we've build/boot an image the CLI will prompt "root@image ~]# "
+            # Then, when pexpects finds the "#" it means we're ready to interact with the process
+            self._serial = pexpect.spawnu(cmd, logfile=None, timeout=240)
+            self._serial.expect("#")
+            self.stack = stack.pop_all()
+
+        return self
+
+
+    def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -> CompletedProcess:
+        self.ensure_booted()
+
+        process = run_command_image(self.args, commands, timeout, check, subprocess.PIPE, subprocess.PIPE)
+        if self.debug:
+            print(f"Stdout:\n {process.stdout}")
+            print(f"Stderr:\n {process.stderr}")
+
+        return process
+
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        if self._serial:
```

Comment:
I'm a bit confused as to what I should do here. If I try:
`if ensure_booted:`
Mypy will point it out:
> mkosi/machine.py:127:13: error: Item "None" of "Optional[Any]" has no attribute "kill"  [union-attr]
                self._serial.kill(signal.SIGTERM)
                ^
> mkosi/machine.py:128:30: error: Item "None" of "Optional[Any]" has no attribute "wait"  [union-attr]
                self.exit_code = self._serial.wait()

Did this only to stop it from failing the CI, but not sure if its good enough

Reply:
[This](https://github.com/systemd/mkosi/pull/894#discussion_r807028122) would solve it then

Reply:
What you currently have is fine.